mlflow/types/schema.py

```diff
@@ -408,3 +408,236 @@ def __eq__(self, other) -> bool:
+    """
+
+    def __init__(
+        self,
+        name: str,
+        type: Union[DataType, str],  # pylint: disable=redefined-builtin
+        default: Union[DataType, List[DataType], None],
+        shape: Optional[tuple] = None,
+    ):
+        self._name = str(name)
+        self._shape = tuple(shape) if shape is not None else None
+
+        try:
+            self._type = DataType[type] if isinstance(type, str) else type
+        except KeyError:
+            raise MlflowException(
+                f"Unsupported type '{type}', expected instance of DataType or "
+                f"one of {[t.name for t in DataType]}"
+            )
+        if not isinstance(self.type, DataType):
+            raise TypeError(
+                "Expected mlflow.models.signature.Datatype or str for the 'type' "
+                f"argument, but got {self.type.__class__}"
+            )
+
+        # This line makes sure repr(self) works fine
+        self._default = default
+        self._default = self.validate_type_and_shape(repr(self), default, self.type, self.shape)
+
+    @classmethod
+    def validate_param_spec(
+        cls, value: Union[DataType, List[DataType], None], param_spec: "ParamSpec"
+    ):
+        return cls.validate_type_and_shape(
+            repr(param_spec), value, param_spec.type, param_spec.shape
+        )
+
+    @classmethod
+    def validate_type_and_shape(
+        cls,
+        spec: str,
+        value: Union[DataType, List[DataType], None],
+        value_type: DataType,
+        shape: Optional[Tuple[int, ...]],
+    ):
+        """
+        Validate that the value has the expected type and shape.
+        """
+
+        def _is_1d_array(value):
+            if not isinstance(value, list):
+                return False
+            if np.array(value).ndim == 1:
+                return True
+            return False
+
+        try:
+            if shape is None:
+                if np.isscalar(value) or value is None:
+                    return (
+                        np.array(value, dtype=value_type.to_numpy()).item()
+                        if value is not None
+                        else None
+                    )
+                else:
+                    raise MlflowException(
+                        f"Shape must be specified for non-scalar value for ParamSpec {spec}",
+                        INVALID_PARAMETER_VALUE,
+                    )
+            elif shape == (-1,):
+                if _is_1d_array(value):
+                    return (
+                        np.array(value, dtype=value_type.to_numpy()).tolist()
+                        if value is not None
+                        else None
+                    )
+                elif np.isscalar(value):
+                    raise MlflowException(
+                        f"Shape must be None for scalar value for ParamSpec {spec}",
+                        INVALID_PARAMETER_VALUE,
+                    )
```

Comment:
This message is a bit confusing at inference time when an invalid shape is passed (e.g. a scalar is passed instead of an array):

```
import mlflow
from mlflow.models import infer_signature

import numpy as np

s = infer_signature(["k"], params={"foo": [8]})


class MyMod(mlflow.pyfunc.PythonModel):
    def predict(self, ctx, inp, params):
        return [7 for item in inp]


with mlflow.start_run():
    m = mlflow.pyfunc.log_model(
        python_model=MyMod(),
        artifact_path="foo",
        signature=s,
        pip_requirements=[],
    )

m = mlflow.pyfunc.load_model(m.model_uri)
print(m.predict(["k"], params={"foo": 8}))
```

```
mlflow.exceptions.MlflowException: Invalid parameters found: {('foo', "Shape must be None for scalar value for ParamSpec 'foo': long (default: [8]) (shape: (-1,))")}
```

I would expect the message to say something like `Found a scalar input, but expected a 1D-array`. Can we update?